File: CSharp/cs_EasyMSX-master/EasyMSX/Brokers.cs
using System.Collections.Generic;

using LogLevels = com.bloomberg.emsx.samples.Log.LogLevels;
using Element = Bloomberglp.Blpapi.Element;
using Message =  Bloomberglp.Blpapi.Message;
using Name = Bloomberglp.Blpapi.Name;
using Request = Bloomberglp.Blpapi.Request;
using System.Collections;

namespace com.bloomberg.emsx.samples {

    public class Brokers : IEnumerable<Broker>  {
	
	    private static readonly Name 	GET_BROKERS = new Name("GetBrokersWithAssetClass");
	    private static readonly Name 	ERROR_INFO = new Name("ErrorInfo");

	    private List<Broker> brokers = new List<Broker>();
        internal EasyMSX emsxapi;
    
	    internal Brokers(EasyMSX emsxapi) {
		    this.emsxapi = emsxapi;
		    loadBrokers();
	    }
	
	    private void loadBrokers() {

		    Log.LogMessage(LogLevels.BASIC,"Brokers: Loading");
	
		    Request reqEQTY = emsxapi.emsxService.CreateRequest(GET_BROKERS.ToString());
		    reqEQTY.Set("EMSX_ASSET_CLASS", Broker.AssetClass.EQTY.ToString());
		    emsxapi.submitRequest(reqEQTY,new BrokersHandler(this, Broker.AssetClass.EQTY));

		    Request reqOPT = emsxapi.emsxService.CreateRequest(GET_BROKERS.ToString());
		    reqOPT.Set("EMSX_ASSET_CLASS", Broker.AssetClass.OPT.ToString());
		    emsxapi.submitRequest(reqOPT,new BrokersHandler(this, Broker.AssetClass.OPT));

		    Request reqFUT = emsxapi.emsxService.CreateRequest(GET_BROKERS.ToString());
		    reqFUT.Set("EMSX_ASSET_CLASS", Broker.AssetClass.FUT.ToString());
		    emsxapi.submitRequest(reqFUT,new BrokersHandler(this, Broker.AssetClass.FUT));

		    Request reqMULTILEFOPT = emsxapi.emsxService.CreateRequest(GET_BROKERS.ToString());
		    reqMULTILEFOPT.Set("EMSX_ASSET_CLASS", Broker.AssetClass.MULTILEG_OPT.ToString());
		    emsxapi.submitRequest(reqMULTILEFOPT,new BrokersHandler(this, Broker.AssetClass.MULTILEG_OPT));
	
	    }

	    class BrokersHandler : MessageHandler {
		
		    Brokers brokers;
		    Broker.AssetClass assetClass;
		
		    internal BrokersHandler(Brokers brokers, Broker.AssetClass assetClass) {
			    this.brokers = brokers;
			    this.assetClass = assetClass;
		    }

		    public void processMessage(Message message) {
			
			    Log.LogMessage(LogLevels.BASIC,"Brokers: processing message");
			
	    	    if(message.MessageType.Equals(ERROR_INFO)) {
	        	    Log.LogMessage(LogLevels.BASIC,"Brokers: processing RESPONSE error");
	    		    int errorCode = message.GetElementAsInt32("ERROR_CODE");
	    		    string errorMessage = message.GetElementAsString("ERROR_MESSAGE");
	    		    Log.LogMessage(LogLevels.BASIC,"Error getting brokers: [" + errorCode + "] " + errorMessage);
	    	    } else if(message.MessageType.Equals(GET_BROKERS)) {
	        	    Log.LogMessage(LogLevels.BASIC,"Brokers: processing successful RESPONSE");
	    		
	        	    Element brokerList = message.GetElement("EMSX_BROKERS");
	    		
				    int numValues = brokerList.NumValues;

				    for(int i = 0; i < numValues; i++) {
	    			
	    			    string brokerName = brokerList.GetValueAsString(i);
	    			    Broker newBroker = new Broker(brokers,brokerName,assetClass);
	    			    brokers.add(newBroker);
	            	    Log.LogMessage(LogLevels.DETAILED,"Brokers: added new broker " + newBroker.name);
	    		    }
	    	    }
		    }
	    }

        public IEnumerator<Broker> GetEnumerator()
        {
            return brokers.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

	    public Broker get(int index) {
		    return brokers[index];
	    }
	
	    public Broker get(string name, Broker.AssetClass assetClass) {
		    foreach(Broker b in brokers){
			    if(b.name.Equals(name) && b.assetClass==assetClass)
			     return b;
		    }
		    return null;
	    }
	
	    public void add(Broker newBroker) {
		    brokers.Add(newBroker);
	    }
    }
}

File: Python/old/py_EasyMSX-master/Brokers.py
# Brokers.py

import blpapi
from Broker import Broker

GET_BROKERS = blpapi.Name("GetBrokersWithAssetClass")
ERROR_INFO = blpapi.Name("ErrorInfo")

class Brokers:
    
    def __init__(self,easyMSX):
        self.easyMSX = easyMSX
        self.brokers=[]
        self.loadBrokers()
    
    def __iter__(self):
        return self.brokers.__iter__()
    
    def loadBrokers(self):
        
        reqEQTY = self.easyMSX.emsxService.createRequest(str(GET_BROKERS))
        reqEQTY.set("EMSX_ASSET_CLASS","EQTY")
        self.easyMSX.submitRequest(reqEQTY, BrokerMessageHandler(self,"EQTY").processMessage)

        reqOPT = self.easyMSX.emsxService.createRequest(str(GET_BROKERS))
        reqOPT.set("EMSX_ASSET_CLASS","OPT")
        self.easyMSX.submitRequest(reqOPT, BrokerMessageHandler(self,"OPT").processMessage)

        reqFUT = self.easyMSX.emsxService.createRequest(str(GET_BROKERS))
        reqFUT.set("EMSX_ASSET_CLASS","FUT")
        self.easyMSX.submitRequest(reqFUT, BrokerMessageHandler(self,"FUT").processMessage)
        
        reqMULTILEG = self.easyMSX.emsxService.createRequest(str(GET_BROKERS))
        reqMULTILEG.set("EMSX_ASSET_CLASS","MULTILEG_OPT")
        self.easyMSX.submitRequest(reqMULTILEG, BrokerMessageHandler(self,"MULTILEG_OPT").processMessage)

class BrokerMessageHandler:
    
    def __init__(self, brokers, assetClass):
        self.brokers = brokers
        self.assetClass = assetClass
        
    def processMessage(self, msg):
        
        if msg.messageType() == ERROR_INFO:
            errorCode = msg.getElementAsInteger("ERROR_CODE")
            errorMessage = msg.getElementAsString("ERROR_MESSAGE")
            print("GetBrokers >> ERROR CODE: %d\tERROR MESSAGE: %s" % (errorCode,errorMessage))

        elif msg.messageType() == GET_BROKERS:
            
            bkrs = msg.getElement("EMSX_BROKERS")
            
            for b in bkrs.values():
                self.brokers.brokers.append(Broker(self.brokers,b,self.assetClass))
                
                                  
__copyright__ = """
Copyright 2017. Bloomberg Finance L.P.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to
deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:  The above
copyright notice and this permission notice shall be included in all copies
or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.
"""
